ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -594,21 +594,44 @@ private void analyzeHaving(final Expression node) {
     }
 
     private void validateSelect(final SingleColumn column) {
-      final ColumnName columnName = column.getAlias()
-          .orElseThrow(IllegalStateException::new);
 
-      if (persistent) {
-        if (SystemColumns.isSystemColumn(columnName, ksqlConfig)) {
-          throw new KsqlException("Reserved column name in select: " + columnName + ". "
-              + "Please remove or alias the column.");
-        }
-      }
+      final int pseudoColumnVersion = SystemColumns.getPseudoColumnVersionFromConfig(ksqlConfig);
+
+      SystemColumns.systemColumnNames(pseudoColumnVersion)
+          .forEach(col -> checkForReservedToken(column, col));
 
       if (!analysis.getGroupBy().isPresent()) {
         throwOnUdafs(column.getExpression());
       }
     }
 
+    private void checkForReservedToken(
+        final SingleColumn singleColumn,
+        final ColumnName reservedToken
+    ) {
+      final ColumnName alias = singleColumn.getAlias().orElseThrow(IllegalStateException::new);
+      final Expression expression = singleColumn.getExpression();
+
+      if (alias.text().equalsIgnoreCase(reservedToken.text())) {
+        if (!expressionMatchesAlias(expression, alias)) {
```

Comment:
Can you add comments to explain both the if and else-if branches of this block? The logic here is not easy for readers to grasp by looking at the code. I'd be confused if not for the fact that we talked about this offline.